packages/extra/src/OffsetDate.js

```diff
@@ -0,0 +1,1274 @@
+/**
+ * @copyright (c) 2016, Philipp Thürwächter & Pattrick Hüper & Michał Sobkiewicz
+ * @copyright (c) 2007-present, Stephen Colebourne & Michael Nascimento Santos
+ * @license BSD-3-Clause (see LICENSE in the root directory of this source tree)
+ */
+
+import { ChronoField, ChronoUnit, Clock, DateTimeException, DateTimeFormatter, IllegalArgumentException, IsoChronology, LocalDate, OffsetDateTime, Temporal, TemporalField, TemporalQueries, TemporalQuery, TemporalUnit, ZoneId, ZoneOffset } from '@js-joda/core';
+
+// TODO: hm... is this a good idea?? copied from joda currently, could we add a js-joda-utils module??
+import { requireInstance, requireNonNull } from './assert';
+import { _ as jodaInternal } from '@js-joda/core';
+
+const MathUtil = jodaInternal.MathUtil;
+
+/**
+ * The number of seconds per day.
+ */
+const SECONDS_PER_DAY = 86400;
+
+/**
+ * A date with an offset from UTC/Greenwich in the ISO-8601 calendar system,
+ * such as `2007-12-03+01:00`.
+ * 
+ * `OffsetDate` is an immutable date-time object that represents a date, often viewed
+ * as year-month-day-offset. This object can also access other date fields such as
+ * day-of-year, day-of-week and week-of-year.
+ * 
+ * This class does not store or represent a time.
+ * For example, the value '2nd October 2007 +02:00' can be stored
+ * in an `OffsetDate`.
+ *
+ */
+export class OffsetDate extends Temporal {
+    //-----------------------------------------------------------------------
+    /**
+     * function overloading for {@link OffsetDate.now}
+     *
+     * if called with 0 argument {@link OffsetDate.now0} is executed,
+     *
+     * if called with 1 argument and first argument is an instance of ZoneId, then {@link OffsetDate.nowZoneId} is executed,
+     *
+     * otherwise {@link OffsetDate.nowClock} is executed
+     *
+     * @param {?(ZoneId|Clock)} zoneIdOrClock
+     * @returns {YearQuarter}
+     */
+    static now(zoneIdOrClock) {
+        if (arguments.length === 0) {
+            return OffsetDate.now0();
+        } else if (arguments.length === 1 && zoneIdOrClock instanceof ZoneId) {
+            return OffsetDate.nowZoneId(zoneIdOrClock);
+        } else {
+            return OffsetDate.nowClock(zoneIdOrClock);
+        }
+    }
+
+    //-----------------------------------------------------------------------
+    /**
+     * Obtains the current date from the system clock in the default time-zone.
+     * 
+     * This will query the {@link Clock.systemDefaultZone} system clock in the default
+     * time-zone to obtain the current date.
+     * The offset will be calculated from the time-zone in the clock.
+     * 
+     * Using this method will prevent the ability to use an alternate clock for testing
+     * because the clock is hard-coded.
+     *
+     * @return {OffsetDate} the current date using the system clock, not null
+     */
+    static now0() {
```

Comment:
why we don't have a leading underscore here ? do we have to add @private annotation, same for the other private methods

Reply:
I used `YearMonth` as reference, however I agree that making this method private would be nice.

Reply:
`Year` and `MonthDay` also have `now0()` with no underscore. Should I fix them too?

Reply:
Yes please as mentioned above, all internal methods should start with an `_` and have the `@private` annotation

Reply:
My impression is that if `now0()` and others becomes private, jsdoc of `now()` is going to be useless. See:
- https://js-joda.github.io/js-joda/class/packages/core/src/Year.js~Year.html#static-method-now
- https://js-joda.github.io/js-joda/class/packages/core/src/YearMonth.js~YearMonth.html#static-method-now
- https://js-joda.github.io/js-joda/class/packages/core/src/MonthDay.js~MonthDay.html#static-method-now

`now()` jsdoc points to `now0()`, `nowClock(clock: Clock)` and `nowZoneId(zone: ZoneId)`. There is similar pattern in case of several `of()` methods. And that's not even the main problem! **These methods are public for at least 6 years**. I'm not sure if making them private now is going to make users happy.

Reply:
I agree with @pithu here. Those methods should be private, and should have been since the beginning. What you are saying about the documentation is true though, but I think that's something that should be fixed at a larger scale. A lot of the documentation needs improvements and the issue you're pointing out is an issue everywhere in this project. Also renaming those methods to start with `_` in other classes (in `core` for instance) is of course a breaking change, but these classes are new; I think is better to make things right from the start here.

However, that being said, I'm an advocate to clean the interfaces in other packages and classes as well, renaming all private methods even if it's a breaking change. Those methods are implementation details and we should be able to change them or delete them at any time.

In this case I would prefer that every method that is not part of the public API is named with an `_` and marked as `@Private`.

Reply:
> Those methods should be private, and should have been since the beginning.

Sure! I do agree with both of you and was even willing (and still am) to go further and use true [private class methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields#private_methods) instead of convention in whole extra package.

> I'm an advocate to clean the interfaces in other packages and classes as well, renaming all private methods even if it's a breaking change.

+1 for cleaning core interfaces sometime later. For now, `@deprecated` could be helpful.

> In this case I would prefer that every method that is not part of the public API is named with an _ and marked as `@Private`.

There are methods (e.g. Temporal's _withField, _withAdjuster, _plusUnit, _plusAmount) which seems to be more `@protected`-like. Still non-public, but with clearly marked purpose.

Reply:
@perceptron8

Using `@deprecated` is a very good suggestion.

> There are methods (e.g. Temporal's _withField, _withAdjuster, _plusUnit, _plusAmount) which seems to be more `@protected`-like. Still non-public, but with clearly marked purpose.

Yes, that's exactly right. We changed (or added) those methods to unify some bits of implementation details and to clean the inheritance chain not so long ago. They are in fact protected in spirit (and in the types definitions), meant to be used by anyone implementing a `Temporal` type. We used `_` as well since they shouldn't be used by anyone outside the inheritance chain. We don't have any established convention right now to differentiate this protected-like methods from private ones, so we used `_` for anything non-public.

About using private fields I personally wouldn't mind. Native support is increasing in every important platform. My only concern is that js-joda still targets ES5, which doesn't include `WeakMap`. How does Babel transpile them to ES5? If it does so by prepending the name with and underscore or something I'd prefer just going as we are for now. Besides, I'm not sure if the performance could degrade if it transpiles to a `WeakMap` in ES6+. 
I think it is worth to consider and discuss the use of private fields/methods, though.

Reply:
I'm afraid that babel transpiles `#` to `WeakMap`, so for now let's stick to `_` convention.

As for ESDoc + overloading related problems, they were solved in likely best possible way thanks to @pithu's https://github.com/js-joda/js-joda/pull/622#issuecomment-1242922282.